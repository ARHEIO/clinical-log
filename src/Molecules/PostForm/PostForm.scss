.post {
  &-submission {
    align-items: center;
    display: flex;
    flex-direction: column;
    margin-top: 16px;

    @media (min-width: 960px) {
      flex-direction: row;
      justify-content: flex-end;
    }
  }
}
