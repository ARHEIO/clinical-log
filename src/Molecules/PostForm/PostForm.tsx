import React, { ReactElement, ReactNode } from 'react';
import Button from '@material-ui/core/Button';
import { Formik } from 'formik';

import './PostForm.scss';
import { TextField, Checkbox, FormControlLabel } from '@material-ui/core';
import Card from '../../Atoms/Card/Card';

const PostForm = (props: {onSubmitParent: Function; isPublic?: boolean}): ReactElement => {
  const { onSubmitParent, isPublic } = props;

  const validate = (values: {postContent: string}): any => {
    const errors: { postContent?: string } = {};
    if (!values.postContent) {
      errors.postContent = 'Post content is required';
    } else if (!isPublic && values.postContent.includes('@')) {
      errors.postContent = 'Mentioned are not allowed in clinical logs';
    }
    return errors;
  };

  return (
    <Card>
      <Formik
        validate={validate}
        initialValues={{ postContent: '', ventCheckbox: false }}
        onSubmit={async (values): Promise<void> => onSubmitParent(values)}
      >
        {(innerProps: any): ReactNode => {
          const {
            values,
            touched,
            errors,
            isSubmitting,
            handleChange,
            handleBlur,
            handleSubmit,
          } = innerProps;
          return (
            <form onSubmit={handleSubmit}>
              <TextField
                id="postContent"
                placeholder="Enter your post"
                type="text"
                value={values.postContent}
                onChange={handleChange}
                variant="outlined"
                onBlur={handleBlur}
                fullWidth
                className="post__input"
              />
              {errors.postContent && touched.postContent && (
                <div className="input-feedback">{errors.postContent}</div>
              )}
              <div className="post-submission">
                {isPublic && (
                  <FormControlLabel
                    className="post-submission__venting"
                    id="ventCheckbox"
                    control={<Checkbox value={values.isVent} onChange={handleChange} id="ventCheckbox" />}
                    label="I'm just venting"
                  />
                )}
                <Button className="post-submission__submit" color="primary" variant="contained" type="submit" disabled={isSubmitting}>
                  Submit
                </Button>
              </div>
            </form>
          );
        }}
      </Formik>
    </Card>
  );
};
export default PostForm;
