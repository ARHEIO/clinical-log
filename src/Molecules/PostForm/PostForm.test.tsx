import React from 'react';
import {
  render, RenderResult, fireEvent, wait,
} from '@testing-library/react';
import pretty from 'pretty';
import PostForm from './PostForm';

const renderComponent = (isPublic: boolean): RenderResult | any => {
  const mockParentFn = jest.fn();
  const result = render(
    <PostForm onSubmitParent={mockParentFn} isPublic={isPublic} />,
  );
  return { ...result, mockParentFn };
};

describe('PostForm', () => {
  it('should render when it is a public component', () => {
    const { container } = renderComponent(true);
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it('should render when it is a private comopnent', () => {
    const { container } = renderComponent(false);
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  describe('submission', () => {
    it('should submit a clinical log', async () => {
      const postContent = 'This is some content to go into the post';
      const { container, mockParentFn } = renderComponent(false);

      fireEvent.change(container.querySelector('#postContent'), { target: { value: postContent } });
      container.querySelector('.post-submission__submit').click();

      await wait(() => {
        expect(mockParentFn).toHaveBeenCalledWith({ postContent });
      });
    });

    it('should submit a public post', async () => {
      const postContent = 'This is some content to go into the post';
      const { container, mockParentFn } = renderComponent(true);

      container.querySelector('input[type="checkbox"]').click();
      fireEvent.change(container.querySelector('#postContent'), { target: { value: postContent } });
      container.querySelector('.post-submission__submit').click();

      await wait(() => {
        expect(mockParentFn).toHaveBeenCalledWith({ postContent });
      });
    });

    it('should show an error if a private log contains an @ symbol', async () => {
      const postContent = 'This is some content to go into the post with a reference to @Adam';
      const { container } = renderComponent(false);

      fireEvent.change(container.querySelector('#postContent'), { target: { value: postContent } });
      container.querySelector('.post-submission__submit').click();

      await wait(() => {
        expect(container.querySelector('.input-feedback').innerHTML).toContain('Mentioned are not allowed in clinical logs');
      });
    });

    it('should chug on if a public post contains an @ symbol', async () => {
      const postContent = 'This is some content to go into the post with a reference to @Adam';
      const { container, mockParentFn } = renderComponent(true);

      fireEvent.change(container.querySelector('#postContent'), { target: { value: postContent } });
      container.querySelector('.post-submission__submit').click();

      await wait(() => {
        expect(mockParentFn).toHaveBeenCalledWith({ postContent });
      });
    });
  });
});
