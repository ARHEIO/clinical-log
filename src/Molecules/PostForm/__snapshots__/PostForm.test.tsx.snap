// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostForm should render when it is a private comopnent 1`] = `
"<div class=\\"card-container\\">
  <form>
    <div class=\\"MuiFormControl-root MuiTextField-root post__input MuiFormControl-fullWidth\\">
      <div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" id=\\"postContent\\" placeholder=\\"Enter your post\\" type=\\"text\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\">
        <fieldset aria-hidden=\\"true\\" style=\\"padding-left: 8px;\\" class=\\"PrivateNotchedOutline-root-187 MuiOutlinedInput-notchedOutline\\">
          <legend class=\\"PrivateNotchedOutline-legend-188\\" style=\\"width: 0.01px;\\"><span>​</span></legend>
        </fieldset>
      </div>
    </div>
    <div class=\\"post-submission\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained post-submission__submit MuiButton-containedPrimary\\" tabindex=\\"0\\" type=\\"submit\\"><span class=\\"MuiButton-label\\">Submit</span><span class=\\"MuiTouchRipple-root\\"></span></button></div>
  </form>
</div>"
`;

exports[`PostForm should render when it is a public component 1`] = `
"<div class=\\"card-container\\">
  <form>
    <div class=\\"MuiFormControl-root MuiTextField-root post__input MuiFormControl-fullWidth\\">
      <div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" id=\\"postContent\\" placeholder=\\"Enter your post\\" type=\\"text\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\">
        <fieldset aria-hidden=\\"true\\" style=\\"padding-left: 8px;\\" class=\\"PrivateNotchedOutline-root-39 MuiOutlinedInput-notchedOutline\\">
          <legend class=\\"PrivateNotchedOutline-legend-40\\" style=\\"width: 0.01px;\\"><span>​</span></legend>
        </fieldset>
      </div>
    </div>
    <div class=\\"post-submission\\"><label class=\\"MuiFormControlLabel-root post-submission__venting\\"><span class=\\"MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-55 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary\\" aria-disabled=\\"false\\"><span class=\\"MuiIconButton-label\\"><input class=\\"PrivateSwitchBase-input-58\\" type=\\"checkbox\\" data-indeterminate=\\"false\\" value=\\"\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></span><span class=\\"MuiTypography-root MuiFormControlLabel-label MuiTypography-body1\\">I'm just venting</span></label><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained post-submission__submit MuiButton-containedPrimary\\" tabindex=\\"0\\" type=\\"submit\\"><span class=\\"MuiButton-label\\">Submit</span><span class=\\"MuiTouchRipple-root\\"></span></button></div>
  </form>
</div>"
`;
